### ISSUE TITLE
AR 5.2.1 and 5.2.2 transitive polymorphic association results in INNER JOIN which breaks "or" searches

### ISSUE THREAD

**Issue by dholdren** [activerecord-hackery/ransack#996]
# Problem
When searching a model's polymorphic association's association's field in Rails 5.2 ransack generates a `INNER JOIN` which causes `or` type searches to break
(i.e. a Note's notable Article's comment's body)

## Working combination: Rails 5.0, ransack 2.1.1 (latest):
This generates `LEFT OUTER JOIN`s for everything
```
➜  ransack git:(master) rake console
[1] pry(main)> puts Note.ransack("notable_of_Article_type_comments_body_cont" => "First").result.to_sql
SELECT "notes".* FROM "notes" 
LEFT OUTER JOIN "articles" ON "articles"."id" = "notes"."notable_id" AND "notes"."notable_type" = 'Article' AND ('default_scope' = 'default_scope') 
LEFT OUTER JOIN "comments" ON "comments"."article_id" = "articles"."id" 
WHERE ("comments"."body" LIKE '%First%')
=> nil

[2] pry(main)> puts Note.ransack("notable_of_Person_type_name_or_notable_of_Article_type_comments_body_cont" => "First").result.to_sql
SELECT "notes".* FROM "notes" 
LEFT OUTER JOIN "people" ON "people"."id" = "notes"."notable_id" AND "notes"."notable_type" = 'Person' 
LEFT OUTER JOIN "articles" ON "articles"."id" = "notes"."notable_id" AND "notes"."notable_type" = 'Article' AND ('default_scope' = 'default_scope') 
LEFT OUTER JOIN "comments" ON "comments"."article_id" = "articles"."id" 
WHERE ("people"."name" LIKE '%First%' OR "comments"."body" LIKE '%First%')
=> nil
```

## Broken combination Rails 5.2.2 (also 5.2.1), ransack 2.1.1 (latest)
This generates an `INNER JOIN` for the `comments` table so it ignores any other tables we chose to search
```
ransack git:(master) RAILS=v5.2.2 rake console
[1] pry(main)> puts Note.ransack("notable_of_Article_type_comments_body_cont" => "First").result.to_sql
SELECT "notes".* FROM "notes" 
LEFT OUTER JOIN "articles" ON "articles"."id" = "notes"."notable_id" AND ('default_scope' = 'default_scope') 
INNER JOIN "comments" ON "comments"."article_id" = "articles"."id" 
WHERE "comments"."body" LIKE '%First%'
=> nil

[2] pry(main)> puts Note.ransack("notable_of_Person_type_name_or_notable_of_Article_type_comments_body_cont" => "First").result.to_sql
SELECT "notes".* FROM "notes" 
LEFT OUTER JOIN "people" ON "people"."id" = "notes"."notable_id" 
LEFT OUTER JOIN "articles" ON "articles"."id" = "notes"."notable_id" AND ('default_scope' = 'default_scope') 
INNER JOIN "comments" ON "comments"."article_id" = "articles"."id" 
WHERE ("people"."name" LIKE '%First%' OR "comments"."body" LIKE '%First%')
=> nil
```

## Additional bug
You can see in the Rails 5.2.2 example, that the polymorphic associations `LEFT OUTER JOIN` aren't limited by `notable_type`. This causes other issues. This was reported in https://github.com/activerecord-hackery/ransack/issues/902

**Comment by dholdren** [comment]
@gregmolnar Is anyone actively looking into this or #955 ? (sorry for the dup)

Also the separate, but related #902 

These are show-stopper bugs for Rails 5.2.1+, polymorphic relationship searches are broken

**Comment by gregmolnar** [comment]
@dholdren I am struggling with time recently and as far as I know others are not looking into this either. If all goes well, I should be able to look into it this weekend/early next week.

**Comment by gregmolnar** [comment]
@dholdren Can you verify that this PR fixes your issue? https://github.com/activerecord-hackery/ransack/pull/1004

**Comment by dholdren** [comment]
> @dholdren Can you verify that this PR fixes your issue? #1004

Yes it looks good. LEFT OUTER joins, and they now have the polymorphic constraint. thanks!
